### PR TITLE
 fixing issues with sporadic failures in UT

### PR DESCRIPTION
**Pull request by jagadish-kb**
1. Caching of deployment names in director client lead to some inconsistencies on the state of cache. Now ensured that before each test run the cache state is the same and if required individual tests can override the default cache state.

2. If there is an error loading cache (i.e. on errors when fetching deployments), then bubble up the error than eating it. Bubbling up the error makes it quite evident where the failures are.

3. If bosh config is being duplicated for a deployment name when fetched from cache, it must throw an error even in this case.

4. During creation of deployment an unnecessary call to BOSH was being fired before picking up the bosh config. Avoided this unnecessary call to bosh during create deployment.

### CODE REVIEW

File: lib/bosh/BoshDirectorClient.js [hunk 1]
```diff
@@ -31,15 +31,16 @@ class BoshDirectorClient extends HttpClient {
     this.oobDirectorConfigs = BoshDirectorClient.getOobDirectorConfigs();
     this.cache = {};
     this.cacheLoadInProgress = false;
-    this.populateCache();
+    this.isReady = this.populateCache();
```
Comment (kamath-prasad) [inline lib/bosh/BoshDirectorClient.js]: `isReady` is a boolean here ?

populateCache() does not return boolean , correct ?

File: lib/bosh/BoshDirectorClient.js [hunk 2]
```diff
@@ -338,29 +339,24 @@ class BoshDirectorClient extends HttpClient {
...
-          logger.warn(`Deploying through ${boshDirectorName}`);
-          return this
-            .getDirectorConfig(deploymentName)
-            .then(config => {
-              if (config) {
-                throw new BadRequest(`Cannot create an existing deployment ${deploymentName}`);
-              }
-            })
-            .catch(NotFound, () => this.getConfigByName(boshDirectorName));
+        if (action === CONST.OPERATION_TYPE.CREATE) {
```
Comment (kamath-prasad) [inline lib/bosh/BoshDirectorClient.js]: 👍

File: lib/bosh/BoshDirectorClient.js (hunk 1 above)
Comment (jagadish-kb) [reply lib/bosh/BoshDirectorClient.js]: its a promise.

Comment (kamath-prasad) [reply lib/bosh/BoshDirectorClient.js]: Ok , should we rename the variable then ? 

As I understand now , it contains the reference to cache wrapped in a promise , correct ?

**Pull request merged**